### PR TITLE
Work on the DB structure a bit

### PR DESCRIPTION
**Pull request by mikeshardmind**
This is still in a very early state, I'll detail it more a little later.

### CODE REVIEW

**Comment by mikeshardmind** [comment]
Still need tables for a few more things:

Automod
Antispam
Warnings
Warn based actions
Scheduled commands
Permission model

Still need efficient queries and Python wrappers

May need to tune the statement cache if we end up above 100 different prepared statements

Once those are in, the design of the command subclass or alternative method for enabling sane xommand scheduling and permission handling can be looked at.

**Comment by mikeshardmind** [comment]
I've added more tables.

I've also started writing the python wrappers.

Apologies in advance for `rachael.sql_wrap.rules_allow_user` @DiscordLiz 

This one's going to need some serious scrutiny, but if we want to allow this to be used publicly, I don't want to be holding all these rules as python objects, so there's a rather complex, but hopefully effcient enough query there.

**Comment by DiscordLiz** [comment]
That's a monstrous query. Are you sure this isn't premature optimization over at least holding the blacklists in memory?

**Comment by mikeshardmind** [comment]
I'm not sure, but this seems like one of those things that doing from the outset will ensure any changes we make are required to stay reasonable on performance and scaling, and not have something ugly creep up and only hit us when scaling kicks in. 

This is the only query that should even approach this level of complexity with the current plans, which is why I wanted to get it written early on with the wrappers. The remaining needed wrappers should all be multiple single-table updates within a transaction or simpler unless I overlooked a planned feature.

I'll try to get this and the 2 other things which were discussed wrapped up soon:tm: so that this can be set up for more collaboration on everything after this, but it won't be tonight, I've been looking at SQL statements for too long.

**Comment by mikeshardmind** [comment]
It's not premature optimization, but I did miss a few things earlier. Full details in the development channel. I'll need to do a lot more with this this week.

**Comment by DiscordLiz** [comment]
Optimizing SQL is squarely outside my domain. I can review this for correctness when you are done, but not for optimal use. We may need to bring on a DBA if you're going to want to continue with some of these ideas beyond what is here or if we ever want to allow more customization of allowed command usage.

**Comment by mikeshardmind** [comment]
I think at the point where enough of this is wrapped and ready to start work, I'll merge this and then add the contributing guidelines we went over to officially open this up. A lot of the tables here won't be used for a while, and that's fine. Part of this was to plan ahead a bit, as long as the table schemas for those themselves are solid, all the queries on them don't have to be ready till that portion of the bot is worked on.

**Comment by DiscordLiz** [comment]
Can you put a pin on the more tables bit for now, and prioritize only the queries needed for the bot core first? The structure of existing tables looks fine. I'll double-check all the foreign key constraints later. I don't know if the queries you have there are optimal, but they work.

**Comment by mikeshardmind** [comment]
Sure. 

I think that's gonna require me to write up a few more, but I'll mark this as ready then rather then when all methods intended are done.

File: rachael/sql_wrap.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        WHERE
+            guild_id = ? AND command = ?
+    ),
+    com_default AS (
+        SELECT false as is_disabled
+    )
+    SELECT * FROM com_vals
+    UNION ALL
+    SELECT * from com_defaults WHERE NOT EXISTS (SELECT * FROM com_vals)
+"""
```
Comment (DiscordLiz) [inline rachael/sql_wrap.py:33]: 1. you forgot a close paren, but I assume this was just a copy-paste error from where you were testing queries based on the indentation.

2. What's the benefit of the two table expressions here (and in most of the other places this is used) over `CASE EXISTS` ?

Comment (DiscordLiz) [inline rachael/sql_wrap.py:33]:
```suggestion
    SELECT * FROM com_vals
    UNION ALL
    SELECT * from com_defaults WHERE NOT EXISTS (SELECT * FROM com_vals)
)
"""
```

Comment (mikeshardmind) [reply rachael/sql_wrap.py]: 1. (applied suggestion, thanks)

2. performance.

To be specific, the query planner won't store the result of the inner select statement.
the select in all of these cases always results in zero or one rows and benefits from an index, which is by design, but by storing the result as table expression it scales better.

**Comment by mikeshardmind** [comment]
There's some more I'm working on locally so that the sqlite interactions all get wrapped in a single class which can simplify some of the way in which this works from the bot side. Haven't settled on a design for that yet, thinking about if this needs to happen in an executor (or more likely a daemon thread with handling to make this play nicer with async) Will do some benchmarking later with the test data we have.

**Comment by mikeshardmind** [comment]
Test data seems to indicate no reason to wrap in an executor at this time. with that in mind, I'll be wrapping this up soon.

**Comment by DiscordLiz** [comment]
Does this need to remain on hold until you finish the other library?

**Comment by mikeshardmind** [comment]
Could probably merge the existing form of this, but since some of it will change with that, I'd rather wait.

**Comment by mikeshardmind** [comment]
Public visibility update:

 - Will be switching over to developing in rust (details pending here)
 - Schema should be mostly the same
 - Currently also considering just using postgresql to simplify some of the concerns, but I'm not sure I want to do that yet.
      - Array types
      - Modifications returning the modified rows

This PR likely won't be finished and any design may be folded into the new project, pending more development discussion about the merits of the 4 approaches being considered.

**Pull request merged**